src/main/flight/pid.c

```diff
@@ -292,13 +315,99 @@ static void pidApplyRateController(const pidProfile_t *pidProfile, pidState_t *p
 #endif
 }
 
+void updateMagHoldHeading(int16_t heading)
+{
+    magHoldTargetHeading = heading;
+}
+
+int16_t getMagHoldHeading() {
+    return magHoldTargetHeading;
+}
+
+uint8_t getMagHoldState()
+{
+
+    #ifndef MAG
+        return MAG_HOLD_DISABLED;
+    #endif
+
+    if (!sensors(SENSOR_MAG) || !STATE(SMALL_ANGLE)) {
+        return MAG_HOLD_DISABLED;
+    }
+
+#if defined(NAV)
+    int navHeadingState = naivationGetHeadingControlState();
+    // NAV will prevent MAG_MODE from activating, but require heading control
+    if (navHeadingState != NAV_HEADING_CONTROL_NONE) {
+        // Apply maghold only if heading control is in auto mode
+        if (navHeadingState == NAV_HEADING_CONTROL_AUTO) {
+            return MAG_HOLD_ENABLED;
+        }
+    }
+    else
+#endif
+    if (ABS(rcCommand[YAW]) < 15 && FLIGHT_MODE(MAG_MODE)) {
```

Comment:
Agreed, it's better to keep this related to actual stick travel, not the mapped rates.